[WIP][Inductor] Enable combo_kernels by default

karthickai opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #170347
* __->__ #168050

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @Lucaskabela @mlazos @chenyang78

karthickai committed:
[WIP][Inductor] Enable combo_kernels by default

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben mlazos

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela mlazos chenyang78

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela mlazos chenyang78

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela mlazos chenyang78

[ghstack-poisoned]

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela mlazos chenyang78

[ghstack-poisoned]

karthickai committed:
Update on "[Inductor] Enable combo_kernels by default"


CI perf on [h100](https://hud.pytorch.org/benchmark/huggingface/inductor_with_cudagraphs?dashboard=torchinductor&startTime=Mon%2C%2001%20Dec%202025%2004%3A54%3A41%20GMT&stopTime=Mon%2C%2008%20Dec%202025%2004%3A54%3A41%20GMT&granularity=hour&mode=training&dtype=amp&deviceName=cuda%20(h100)&lBranch=main&lCommit=c55e1557a9a748628d1cf5672ccc9c508c0199b6&rBranch=gh/karthickai/14/head&rCommit=b2722288f0bfacc17d6d58ed3df3967d994d0cba) getting good speedups on training transformer models by reducing kernel launch overhead and improving GPU utilization.

- 1.2x average speedup across 27 HuggingFace models
- 1.42x speedup (GPT2ForSequenceClassification: 17.90ms → 12.59ms)
- Minimal compilation overhead: Typically +0-3 seconds compile time

[
<img width="1669" height="958" alt="Screenshot 2025-12-08 at 5 40 09 AM" src="https://github.com/user-attachments/assets/23fed150-5177-4a6f-b6f8-85025d4cf680" />
](url)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela mlazos chenyang78

[ghstack-poisoned]

karthickai committed:
Update on "[Inductor] Enable combo_kernels by default"


CI perf on [h100](https://hud.pytorch.org/benchmark/huggingface/inductor_with_cudagraphs?dashboard=torchinductor&startTime=Mon%2C%2001%20Dec%202025%2004%3A54%3A41%20GMT&stopTime=Mon%2C%2008%20Dec%202025%2004%3A54%3A41%20GMT&granularity=hour&mode=training&dtype=amp&deviceName=cuda%20(h100)&lBranch=main&lCommit=c55e1557a9a748628d1cf5672ccc9c508c0199b6&rBranch=gh/karthickai/14/head&rCommit=b2722288f0bfacc17d6d58ed3df3967d994d0cba) getting good speedups on training transformer models by reducing kernel launch overhead and improving GPU utilization.

- 1.2x average speedup across 27 HuggingFace models
- 1.42x speedup (GPT2ForSequenceClassification: 17.90ms → 12.59ms)
- Minimal compilation overhead: Typically +0-3 seconds compile time

[
<img width="1669" height="958" alt="Screenshot 2025-12-08 at 5 40 09 AM" src="https://github.com/user-attachments/assets/23fed150-5177-4a6f-b6f8-85025d4cf680" />
](url)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela mlazos chenyang78

[ghstack-poisoned]

karthickai committed:
Update on "[Inductor] Enable combo_kernels by default"


CI perf on [h100](https://hud.pytorch.org/benchmark/huggingface/inductor_with_cudagraphs?dashboard=torchinductor&startTime=Mon%2C%2001%20Dec%202025%2004%3A54%3A41%20GMT&stopTime=Mon%2C%2008%20Dec%202025%2004%3A54%3A41%20GMT&granularity=hour&mode=training&dtype=amp&deviceName=cuda%20(h100)&lBranch=main&lCommit=c55e1557a9a748628d1cf5672ccc9c508c0199b6&rBranch=gh/karthickai/14/head&rCommit=b2722288f0bfacc17d6d58ed3df3967d994d0cba) getting good speedups on training transformer models by reducing kernel launch overhead and improving GPU utilization.

- 1.2x average speedup across 27 HuggingFace models
- 1.42x speedup (GPT2ForSequenceClassification: 17.90ms → 12.59ms)
- Minimal compilation overhead: Typically +0-3 seconds compile time

[
<img width="1669" height="958" alt="Screenshot 2025-12-08 at 5 40 09 AM" src="https://github.com/user-attachments/assets/23fed150-5177-4a6f-b6f8-85025d4cf680" />
](url)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela mlazos chenyang78

[ghstack-poisoned]

eellison commented:
@jataylo yea we plan on investigating slowdowns before landing (on both cuda/rocm).

karthickai committed:
Update on "[Inductor] Enable combo_kernels by default"


CI perf on [h100](https://hud.pytorch.org/benchmark/huggingface/inductor_with_cudagraphs?dashboard=torchinductor&startTime=Mon%2C%2001%20Dec%202025%2004%3A54%3A41%20GMT&stopTime=Mon%2C%2008%20Dec%202025%2004%3A54%3A41%20GMT&granularity=hour&mode=training&dtype=amp&deviceName=cuda%20(h100)&lBranch=main&lCommit=c55e1557a9a748628d1cf5672ccc9c508c0199b6&rBranch=gh/karthickai/14/head&rCommit=b2722288f0bfacc17d6d58ed3df3967d994d0cba) getting good speedups on training transformer models by reducing kernel launch overhead and improving GPU utilization.

- 1.2x average speedup across 27 HuggingFace models
- 1.42x speedup (GPT2ForSequenceClassification: 17.90ms → 12.59ms)
- Minimal compilation overhead: Typically +0-3 seconds compile time

[
<img width="1669" height="958" alt="Screenshot 2025-12-08 at 5 40 09 AM" src="https://github.com/user-attachments/assets/23fed150-5177-4a6f-b6f8-85025d4cf680" />
](url)


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela mlazos chenyang78

[ghstack-poisoned]

karthickai commented on test/inductor/test_flex_attention.py:
```diff
@@ -1398,6 +1398,8 @@ def test_kv_batch_broadcast(
     @supported_platform
     @skip_on_cpu
+    # dynamic shape on reduction dimension is not supported in combo kernels
+    @config.patch("combo_kernels", False)
```
if we dont patch we will get this `Dynamic shape on reduction dimension is not supported` Runtime error.

```bash
======================================================================
ERROR: test_small_block_mask_cuda (__main__.TestFlexAttentionCUDA.test_small_block_mask_cuda)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/karthickps/combo_kernels/torch/testing/_internal/common_utils.py", line 3329, in wrapper
    method(*args, **kwargs)
    ~~~~~~^^^^^^^^^^^^^^^^^
  File "/home/karthickps/combo_kernels/torch/testing/_internal/common_utils.py", line 3329, in wrapper
    method(*args, **kwargs)
    ~~~~~~^^^^^^^^^^^^^^^^^
  File "/home/karthickps/combo_kernels/torch/testing/_internal/common_device_type.py", line 448, in instantiated_test
    raise rte
  File "/home/karthickps/combo_kernels/torch/testing/_internal/common_device_type.py", line 428, in instantiated_test
    result = test(self, **param_kwargs)
  File "/home/karthickps/combo_kernels/torch/testing/_internal/common_device_type.py", line 1381, in dep_fn
    return fn(self, *args, **kwargs)
  File "/home/karthickps/combo_kernels/torch/testing/_internal/common_device_type.py", line 1231, in dep_fn
    return fn(slf, *args, **kwargs)
  File "/home/karthickps/combo_kernels/test/inductor/test_flex_attention.py", line 1443, in test_small_block_mask
    create_block_mask_from_seqlens(seqlen, seqlen)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^
  File "/home/karthickps/combo_kernels/test/inductor/test_flex_attention.py", line 1430, in create_block_mask_from_seqlens
    return compiled_create_block_mask(
        batch_mask_mod,
    ...<4 lines>...
        device=device,
    )
  File "/home/karthickps/combo_kernels/torch/_dynamo/eval_frame.py", line 956, in compile_wrapper
    raise e.remove_dynamo_frames() from None  # see TORCHDYNAMO_VERBOSE=1
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/karthickps/combo_kernels/torch/_inductor/compile_fx.py", line 1019, in _compile_fx_inner
    raise InductorError(e, currentframe()).with_traceback(
        e.__traceback__
    ) from None
  File "/home/karthickps/combo_kernels/torch/_inductor/compile_fx.py", line 1003, in _compile_fx_inner
    mb_compiled_graph = fx_codegen_and_compile(
        gm, example_inputs, inputs_to_check, **graph_kwargs
    )
  File "/home/karthickps/combo_kernels/torch/_inductor/compile_fx.py", line 1757, in fx_codegen_and_compile
    return scheme.codegen_and_compile(gm, example_inputs, inputs_to_check, graph_kwargs)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/karthickps/combo_kernels/torch/_inductor/compile_fx.py", line 1537, in codegen_and_compile
    compiled_module = graph.compile_to_module()
  File "/home/karthickps/combo_kernels/torch/_inductor/graph.py", line 2416, in compile_to_module
    return self._compile_to_module()
           ~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/home/karthickps/combo_kernels/torch/_inductor/graph.py", line 2422, in _compile_to_module
    self.codegen_with_cpp_wrapper() if self.cpp_wrapper else self.codegen()
                                                             ~~~~~~~~~~~~^^
  File "/home/karthickps/combo_kernels/torch/_inductor/graph.py", line 2358, in codegen
    self.scheduler.codegen()
    ~~~~~~~~~~~~~~~~~~~~~~^^
  File "/home/karthickps/combo_kernels/torch/_inductor/scheduler.py", line 6037, in codegen
    self._codegen_partitions()
    ~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/home/karthickps/combo_kernels/torch/_inductor/scheduler.py", line 6178, in _codegen_partitions
    self._codegen(partition)
    ~~~~~~~~~~~~~^^^^^^^^^^^
  File "/home/karthickps/combo_kernels/torch/_inductor/scheduler.py", line 6277, in _codegen
    backend.codegen_combo_kernel(node)
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/home/karthickps/combo_kernels/torch/_inductor/codegen/cuda_combined_scheduling.py", line 139, in codegen_combo_kernel
    return self._triton_scheduling.codegen_combo_kernel(*args, **kwargs)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
  File "/home/karthickps/combo_kernels/torch/_inductor/codegen/simd.py", line 2341, in codegen_combo_kernel
    kernel_code_list = self.generate_combo_kernel_code(
        subkernel_nodes, custom_part_algorithm, enable_autotune, mixed_sizes
    )
  File "/home/karthickps/combo_kernels/torch/_inductor/codegen/simd.py", line 2329, in generate_combo_kernel_code
    src_code = kernel.codegen_kernel()
  File "/home/karthickps/combo_kernels/torch/_inductor/codegen/triton_combo_kernel.py", line 827, in codegen_kernel
    uniquify = self.codegen_static_numels_sub_kernel(
        code, sub_kernel, num
    )
  File "/home/karthickps/combo_kernels/torch/_inductor/codegen/triton_combo_kernel.py", line 450, in codegen_static_numels_sub_kernel
    raise RuntimeError(
        "Dynamic shape on reduction dimension is not supported"
    )
torch._inductor.exc.InductorError: RuntimeError: Dynamic shape on reduction dimension is not supported

Set TORCHDYNAMO_VERBOSE=1 for the internal stack trace (please do this especially if you're reporting a bug to PyTorch). For even more developer context, set TORCH_LOGS="+dynamo"


To execute this test, run the following from the base repo dir:
    python test/inductor/test_flex_attention.py TestFlexAttentionCUDA.test_small_block_mask_cuda

This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0

----------------------------------------------------------------------
Ran 2 tests in 45.315s

FAILED (errors=1, skipped=1)
```

drisspg replied:
If we are planning to enable this by default this feels bad.. or do we have plans to resolve before land?

karthickai committed:
Update on "[WIP][Inductor] Enable combo_kernels by default"


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy kadeng muchulee8 amjames chauhang aakhundov coconutruben jataylo Lucaskabela mlazos chenyang78

[ghstack-poisoned]
